src/server.c

```diff
@@ -6690,7 +6690,7 @@ void loadDataFromDisk(void) {
         rdbSaveInfo rsi = RDB_SAVE_INFO_INIT;
         errno = 0; /* Prevent a stale value from affecting error checking */
         int rdb_flags = RDBFLAGS_NONE;
-        if (iAmMaster()) {
+        if (iAmMaster() && access(server.rdb_filename, F_OK) == 0) {
```

Comment:
why did you add that?
is it because you wanna revert the creation of the backlog if rdb load fails or rdb file is missing?
i'm afraid that `access` and `open` will give different results.
or actually, we do have a plan to one day allow a module to provide an alternative to file system access.
i'll feel better if we can solve this differently, like destroy the backlog if no loading happened, or create it using some hook

Reply:
Because it makes maxmemory tests failed, current solution for replication backlog(blocks list) makes its exceed our setting.
And i think this change makes sense. if there is no rdb, we don't need to create backlog.
if rdb load fails, redis will exits. we allow redis run only if there is no rdb, of course, we can free replication backlog if errno is ENOENT later.

Reply:
so i'd rather free it laster than use a different method to check if it exists ahead of time.

however, i didn't understand which test failed and why.. can you elaborate?

Reply:
and which test did it break? which assertion?

Reply:
@oranagra i make a mistakes, ignore my analysis above
we will get OOM error in this line  https://github.com/redis/redis/blob/unstable/tests/unit/maxmemory.tcl#L189 
i add `puts "db9 [s db9] [s evicted_keys] [s mem_replication_backlog]"` before this line, i got this result
```
b9 keys=84,expires=84,avg_ttl=9999950 591 101524
db9 keys=85,expires=85,avg_ttl=9999950 591 101524
db9 keys=84,expires=84,avg_ttl=9999950 593 101524
db9  678 118444
```
in this test, we want to write about 96k data, but now replication backlog grows incrementally, i guess we only write 48k, and replication backlog also about 48k, and then we want to write the same number keys,  before we set maxmemory to `used_memory+100k`, so eviction would happen, replication backlog will be bigger, replication backlog is 1M by default, i don't count it inot `NotCountedMemory`, so server evicts all keys.

Reply:
ok. one could argue that this test is over sensitive. if the test was using a replicated pair of servers, we would have had to change the test.
but since the test isn't using replication, and isn't loading an RDB (like many others), let's try to make the memory usage more predictable for them.
i.e. in the past (before Soloestoy's recent change), there was on replication backlog in these tests.
then the backlog was added, but its memory consumption was constant, so this test didn't notice.
now its memory consumption gradually grows, and that breaks the test.
let's make sure to avoid creating the backlog in that case, or create it and then destroy it if not needed.

but i don't like the current solution.
the better one IMHO is that if we didn't set `server.master_repl_offset` from `rsi.repl_offset` (or didn't load an rdb file at all), then we should destroy the backlog.

Reply:
I updated commits, but it seems i forget to handle invalid `rsi.repl_offset`

Reply:
what do you mean?
p.s. an alternative approach to what you took, is that instead of integrating the release of the backlog in the if-else chains of the loading, you can just add an `if (server.master_repl_offset == 0)` at the very end.